### PR TITLE
Horovod grad agg

### PR DESCRIPTION
**Pull request by brainhart**

### CODE REVIEW

File: examples/FasterRCNN/modeling/generalized_rcnn.py [hunk 1]
```diff
@@ -40,9 +40,9 @@ def optimizer(self):
 
         # The learning rate in the config is set for 8 GPUs, and we use trainers with average=False.
         lr = lr / 8.
-        opt = tf.train.MomentumOptimizer(lr, 0.9)
-        if cfg.TRAIN.NUM_GPUS < 8:
-            opt = optimizer.AccumGradOptimizer(opt, 8 // cfg.TRAIN.NUM_GPUS)
+        opt = tf.train.MomentumOptimizer(lr, 0)
+#        if cfg.TRAIN.NUM_GPUS < 8:
+#            opt = optimizer.AccumGradOptimizer(opt, 8 // cfg.TRAIN.NUM_GPUS)
```
Comment (aaron276h) [inline examples/FasterRCNN/modeling/generalized_rcnn.py]: Lets comment these back in.

File: examples/FasterRCNN/train.py [hunk 2]
```diff
@@ -119,7 +119,7 @@
         starting_epoch=cfg.TRAIN.STARTING_EPOCH
     )
     if is_horovod:
-        trainer = HorovodTrainer(average=False)
+        trainer = HorovodTrainer(average=False, aggregation_frequency=2)
```
Comment (aaron276h) [inline examples/FasterRCNN/train.py:122]: Not a big need, but maybe make this be read from the config file.

File: tensorpack/train/trainers.py [hunk 3]
```diff
@@ -403,23 +416,88 @@ def allreduce(self, grads):
...
-            opt = get_opt_fn()
-            self.train_op = opt.apply_gradients(grads, name='train_op')
-            self.comm_op = tf.no_op()
+            self.grads = self._make_get_grad_fn(input, get_cost_fn, get_opt_fn)()
+            aggregation_ops_list = []
+            if self._aggregation_frequency > 1:
+                # Create the variables which will be used to store aggregated updates.
+                with tf.variable_scope("aggregation_variables"):
+                    for idx, (grad, var) in enumerate(self.grads):
+                        grad_aggregation_variable_name = str(idx)
```
Comment (aaron276h) [inline tensorpack/train/trainers.py:440]: Could be useful for a comment here about why it's ok to use the idx for the name versus what we had done previously.

**Review by aaron276h — APPROVED** [review]

File: examples/FasterRCNN/modeling/generalized_rcnn.py (hunk 1 above)
Comment (aaron276h) [reply examples/FasterRCNN/modeling/generalized_rcnn.py]: on second thought lets just delete this.

File: examples/FasterRCNN/train.py (hunk 2 above)
Comment (brainhart) [reply examples/FasterRCNN/train.py:122]: gonna hold off for now, should be easy enough to do later if we want

**Pull request merged**